File: bindings/python/samples/pair_callback_sample.py
import threading

import zlink
from sample_support import tcp_endpoint, wait_connected


def main():
    _, endpoint = tcp_endpoint()

    with zlink.Context() as ctx:
        with zlink.PairSocket(ctx) as server:
            with zlink.PairSocket(ctx) as client:
                with server.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as server_monitor:
                    with client.monitor_open(zlink.MonitorEventMask.CONNECTION_READY) as client_monitor:
                        server.bind(endpoint)
                        client.connect(endpoint)
                        wait_connected(server_monitor, client_monitor)

                done = threading.Event()
                observed = {}

                def on_ready(_sock):
                    done.set()

                server.on_send_ready(on_ready)
                client.send(b"hello-pair")
                if not done.wait(3.0):
                    raise TimeoutError("pair send-ready callback did not fire")
                with server.recv() as received:
                    observed["payload"] = received.to_bytes_list()
                if observed["payload"] != [b"hello-pair"]:
                    raise AssertionError(f"unexpected pair payload: {observed['payload']!r}")
                print('[pair/callback] send-ready + recv: "hello-pair"')


if __name__ == "__main__":
    main()

File: bindings/node/samples/pair_callback_sample.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../dist/canonical');

async function reservePort() {
  const srv = net.createServer();
  srv.listen(0, '127.0.0.1');
  await once(srv, 'listening');
  const { port } = srv.address();
  await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
  const port = await reservePort();
  const endpoint = `tcp://127.0.0.1:${port}`;
  const ctx = new zlink.Context();
  const server = new zlink.PairSocket(ctx);
  const client = new zlink.PairSocket(ctx);

  try {
    const serverMonitor = server.monitorOpen(zlink.MonitorEvent.CONNECTION_READY);
    const clientMonitor = client.monitorOpen(zlink.MonitorEvent.CONNECTION_READY);
    try {
      server.bind(endpoint);
      client.connect(endpoint);
      serverMonitor.recv();
      clientMonitor.recv();
    } finally {
      serverMonitor.close();
      clientMonitor.close();
    }

    const received = await new Promise((resolve, reject) => {
      try {
        const deadline = Date.now() + 5000;
        const poll = () => {
          try {
            resolve(server.recv(zlink.RecvFlags.DontWait));
            return;
          } catch (error) {
            if (
              error instanceof zlink.RecvError
              && error.result === zlink.RecvResult.NoData
            ) {
              if (Date.now() >= deadline) {
                reject(new Error('pair callback sample timed out'));
                return;
              }
              setImmediate(poll);
              return;
            }
            reject(error);
          }
        };
        poll();
      } catch (error) {
        reject(error);
        return;
      }
      const sent = 'hello-pair';
      client.send(Buffer.from(sent));
    });

    assert.ok(received.routingId instanceof zlink.RoutingId);
    const recv = received.parts[0].data().toString();
    assert.equal(recv, 'hello-pair');
    console.log(`[pair/callback] send: "hello-pair" \u2192 recv: "${recv}"`);
  } finally {
    client.close();
    server.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
